### PR TITLE
Support writing/reading notes to/from excel files 

### LINKED ISSUE #58070: ENH: styler.to_excel to support excel notes as implementation of tooltips
### Feature Type

- [X] Adding new functionality to pandas

- [ ] Changing existing functionality in pandas

- [ ] Removing existing functionality in pandas

### Problem Description

I wish I could use set_tooltips in the styler to generate notes in an excel workbook (via the to_excel)

### Feature Description

No real API change, but have the ExcelFormatter use the information set by styler.set_tooltips to create not on the cells with a tooltip

### Alternative Solutions

No idea for an alternative except the statu quo...

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by diogomsmiranda**
- [x] closes #58070
- [x] [Tests added and passed]
- [x] All [code checks passed]
- [x] Added [type annotations]
- [x] Added an entry in the latest `doc/source/whatsnew/v3.0.0.rst` file if fixing a bug or adding a new feature.

Me and my colleague @Dacops developed this feature with the use case issued in mind.
We believe the solution can be better understood when looking at the examples we provide in the documentation but also in the tests we made.

Unfortunately, not every engine supported by pandas had the capability to handle notes (calamine, odfreader, pyxlsb, odswriter), on the remaining ones we took the liberty to not only add writing capability as requested in #58070 but also add the complementary operation of reading notes.

In summary, now  you can write notes to an excel (using either openpyxl or xlswriter) by doing:
 ```
 df = pandas.DataFrame(...) # DataFrame holding cell content
 df_notes = pandas.DataFrame(...)  # DataFrame holding notes
 df.style.set_tooltips(df_notes).to_excel(excel_file.xlsx)
```
 The complementary of this (reading using either xlrd or openpyxl) would look something like:
 ```
df_notes = pandas.DataFrame() # empty DataFrame
pandas.read_excel(path, engine="xlrd", notes=df_notes) # df_notes now hold the notes as they were in the excel
```

We will be available and ready to answer any doubts related to the pr.

Hope everything is to your liking.

### CODE REVIEW

**Comment by attack68** [comment]
I think this looks like a good PR overall but this is not actually what I originally suggested. The `ExcelFormatter` class in pandas is capable of detecting a `Styler` object and what I suggested was that that branching could be extended to handle an attached Styler tooltips object. I.e. I was proposing modifying the `Styler.to_excel` method and not the core `DataFrame.to_excel` method.

However, what you have built is the generic ability for the `to_excel` method to handle comments in the forms of `notes` which is an additional keyword argument to the underlying function. Additionally when looking at first glance you seem to have added the ability to read notes as well. This may well all be quite useful functionality.

So this is a comprehensive PR. It may have been best to segregate these different components into different successive PRs, to make it easier to properly review and consider all the edge cases.

One case that springs to mind is that the way tooltips works for Styler is to `reindex_like` so that its columns and index align with the underlying dataframe. This does not appear to be the case for `notes`, so some kind of consideration for alignment, or properly documenting the difference would probably be needed.

Before doing a full review it will be interesting to hear from other core devs. Also in regards to potential spilt of the PR.

**Comment by WillAyd** [comment]
I am -.5 on adding this as a keyword argument to the excel I/O functions; the fact that so few of the underlying libraries supports this is a red flag. Generally we have a really high bar to expand our API, and adding this without supporting all underlying libraries is confusing for end users

**Comment by diogomsmiranda** [comment]
> I think this looks like a good PR overall but this is not actually what I originally suggested. The `ExcelFormatter` class in pandas is capable of detecting a `Styler` object and what I suggested was that that branching could be extended to handle an attached Styler tooltips object. I.e. I was proposing modifying the `Styler.to_excel` method and not the core `DataFrame.to_excel` method.

@attack68 Thank you for the comment, and sorry for our miss interpretation, I've adjusted the description of the PR accordingly.

> One case that springs to mind is that the way tooltips works for Styler is to `reindex_like` so that its columns and index align with the underlying dataframe. This does not appear to be the case for `notes`, so some kind of consideration for alignment, or properly documenting the difference would probably be needed.

That's indeed a point we noticed once we were testing the code, however, we didn't know what would be the right way to approach this, we'll await further instructions from core developers as you suggested.

**Comment by rhshadrach** [comment]
> > I.e. I was proposing modifying the `Styler.to_excel` method and not the core `DataFrame.to_excel` method.
> 
> @attack68 Thank you for the comment, and sorry for our miss interpretation, I've adjusted the description of the PR accordingly.

But this PR is still modifying the core `DataFrame.to_excel` method?

> However, after a detailed examination of the source code from the relevant libraries, we discovered that many are no longer actively developed and/or have limited functionality, `python-calamine`, however, is exception to this as it's relatively new still. This likely contributes to the current lack of support for notes.

While `python-calamine` is new, the Rust library it uses, `calamine` has been around since 2016. And in the ReadMe, they state

> Many (most) part of the specifications are not implemented, the focus has been put on reading cell values and vba code.

From this, it seems unlikely for calamine to gain the ability to read notes. If I'm understanding this PR correctly, only openpyxl and xlrd support reading comments. Coupled with the fact that the API is, in my opinion, somewhat awkward (passing in a DataFrame to store the result), I am rather negative on supporting reading notes as well.

Somewhat of a separate question - are there engines which are no longer maintained and they provide no additional functionality from libraries that are maintained? These could be deprecated - xlrd comes to mind, but I'm not sure if calamine completely replaces it.

**Comment by Dacops** [comment]
> But this PR is still modifying the core `DataFrame.to_excel` method?

No, **to_excel** is not modified, the notes are kept as an attribute of **ExcelFormatter** (defaulted to None) that is then received by the writers. 

> From this, it seems unlikely for calamine to gain the ability to read notes. If I'm understanding this PR correctly, only openpyxl and xlrd support reading comments. Coupled with the fact that the API is, in my opinion, somewhat awkward (passing in a DataFrame to store the result), I am rather negative on supporting reading notes as well.

Reading notes was initially added for testing purposes as an afterthought. It's a more intrusive method and is supported by fewer libraries. Now that writing notes has demonstrated its intended functionality, this feature could be removed.

> Somewhat of a separate question - are there engines which are no longer maintained and they provide no additional functionality from libraries that are maintained? These could be deprecated - xlrd comes to mind, but I'm not sure if calamine completely replaces it.

Based on our research, the writers xlswriter and openpyxl both write to .xlsx files (I believe Pandas uses both in parallel due to some compatibility issues). The odswriter writes to .ods files.
For the readers, xlrd reads from old-style Excel formats (.xls), while openpyxl reads from .xlsx, .xlsm, .xltx, and .xltm files. Odfreader handles OpenDocument type files (.odf, .ods, .odt), and pyxlsb reads binary (.xlsb) files.
Calamine is the most interesting package as it reads from both Excel (.xls, .xlsx, .xlsm, .xlsb) and OpenDocument (.ods) formats. It is also a newer package. I believe some readers could be replaced by Calamine; it’s just a matter of checking for any compatibility issues.

**Comment by Dacops** [comment]
Rebased to address a conflict with this [commit](https://github.com/pandas-dev/pandas/pull/58921)

**Comment by rhshadrach** [comment]
> Reading notes was initially added for testing purposes as an afterthought. It's a more intrusive method and is supported by fewer libraries. Now that writing notes has demonstrated its intended functionality, this feature could be removed.

@attack68 - you're good with just writing notes instead of having reading as well?

**Comment by attack68** [comment]
Yes that sounds good. I think the writing mode is a good feature and relatively speaking it is a fairly neglible extension to the overall API. I think this should be accepted to merge.

I understand that the reading was added probably as a 'round trip' test whcih is sensible but reading is a separate process and this could (and should) be added in a separate, atomic PR, albeit that seems like lesser chance of acceptance.

**Comment by diogomsmiranda** [comment]
> Yes that sounds good. I think the writing mode is a good feature and relatively speaking it is a fairly neglible extension to the overall API. I think this should be accepted to merge.
> 
> I understand that the reading was added probably as a 'round trip' test whcih is sensible but reading is a separate process and this could (and should) be added in a separate, atomic PR, albeit that seems like lesser chance of acceptance.

We are also ok with this. @rhshadrach should we now delete the reading from this pr and open a separate one?

**Comment by rhshadrach** [comment]
@diogomsmiranda - modifying this PR would be preferable.

**Comment by Dacops** [comment]
This should already have only the writing notes part, just need to figure out why is that docstring check failing.

**Comment by Dacops** [comment]
@rhshadrach We've modified the PR, however we are having a weird issue with the docstrings validation, any idea what might be causing this problem?

File: pandas/io/formats/style.py [hunk 1]
```diff
@@ -537,6 +538,19 @@ def set_tooltips(
         storage_options=_shared_docs["storage_options"],
         storage_options_versionadded="1.5.0",
         extra_parameters="",
+        extra_examples=textwrap.dedent(
+            """\
```
Comment (WillAyd) [inline pandas/io/formats/style.py:542]:
```suggestion
            """
```

Maybe this is causing your docstring error - is this a reason you needed to add this?

File: pandas/core/generic.py [hunk 2]
```diff
@@ -2127,10 +2127,11 @@ def _repr_data_resource_(self):
         storage_options_versionadded="1.2.0",
         extra_parameters=textwrap.dedent(
             """\
-        engine_kwargs : dict, optional
-            Arbitrary keyword arguments passed to excel engine.
-    """
+            engine_kwargs : dict, optional
```
Comment (WillAyd) [inline pandas/core/generic.py:2130]: Did the indentation need to change on this? Also likely another reason for the docstring failures

File: pandas/io/formats/style.py (hunk 1 above)
Comment (diogomsmiranda) [reply pandas/io/formats/style.py:542]: We wrote it like this because it was present in other docstrings, so we followed that example, I'll add a commit with the change and see If the error stands, thanks.

File: pandas/core/generic.py (hunk 2 above)
Comment (diogomsmiranda) [reply pandas/core/generic.py:2130]: I agree, this was probably something that went unnoticed, I'll fix it.

File: pandas/io/excel/_openpyxl.py [hunk 3]
```diff
@@ -484,7 +489,11 @@ def _write_cells(
                 row=freeze_panes[0] + 1, column=freeze_panes[1] + 1
             )
 
+        notes_col = None
+
         for cell in cells:
+            if notes_col is None:
+                notes_col = startcol + cell.col + 1
```
Comment (rhshadrach) [inline pandas/io/excel/_openpyxl.py:496]: Can you instead do e.g. `next(iter(cells)).col` and take this out of the for loop?

File: pandas/io/excel/_openpyxl.py [hunk 4]
```diff
@@ -530,6 +539,20 @@ def _write_cells(
                             for k, v in style_kwargs.items():
                                 setattr(xcell, k, v)
 
+        if notes is None or notes_col is None:
+            return
+
+        for row_idx, (_, val) in enumerate(notes.iterrows()):
```
Comment (rhshadrach) [inline pandas/io/excel/_openpyxl.py:545]: iterrows creates a Series per row, which is very expensive. Use `itertuples(index=False)` instead.

File: pandas/io/excel/_openpyxl.py [hunk 5]
```diff
@@ -530,6 +539,20 @@ def _write_cells(
...
+            for col_idx, note in enumerate(val):
+                xcell = wks.cell(
+                    # first row has columns and openpyxl starts counting at 1, not 0
+                    row=row_idx + 2,
+                    column=col_idx + notes_col,  # n columns with indexes
+                )
+                if note:
```
Comment (rhshadrach) [inline pandas/io/excel/_openpyxl.py:552]: Is `note` here guaranteed to be a string?

**Review by rhshadrach — CHANGES_REQUESTED** [review]
I put remarks on one of the writers - I think this applies to all writers. This also needs tests. While pandas does not have the ability to read the notes, you can use one of the Excel engines that does to read them directly.

What happens in the case of merged cells?

File: pandas/io/excel/_openpyxl.py (hunk 5 above)
Comment (Dacops) [reply pandas/io/excel/_openpyxl.py:552]: indeed it is not, that was casted in the _xlsxwriter engine but was missing here

File: pandas/io/excel/_openpyxl.py (hunk 4 above)
Comment (Dacops) [reply pandas/io/excel/_openpyxl.py:545]: Thanks for this, I've replaced it and it simplifies the iterating even more now it's just `val` instead of `(_, val)`.

File: pandas/io/excel/_openpyxl.py (hunk 3 above)
Comment (Dacops) [reply pandas/io/excel/_openpyxl.py:496]: I'm not getting this one, `notes_col` is the 1st column where data is inserted (to skip over all index columns) since when writing to excel I need the actual cell column number (in the excel sheet).

**Comment by Dacops** [comment]
> I put remarks on one of the writers - I think this applies to all writers. 
I've updated this in all writers, except the first example, which I didn't understand, could you better explain how do you want us to change that logic?

> This also needs tests. While pandas does not have the ability to read the notes, you can use one of the Excel engines that does to read them directly.
I've brought all the tests from the original issue except now the reading is handled directly by the library `openpyxl` in the testing file.

> What happens in the case of merged cells?
The comment writing occurs after all the content is written, so if the notes are passed in mind with cell merging that's how the writing will proceed.


Also some `datetime` tests started failing on the 32-bits test suites due to overflows, is this a general problem?

File: pandas/io/excel/_odswriter.py [hunk 6]
```diff
@@ -110,6 +113,14 @@ def _write_cells(
         )
         from odf.text import P
 
+        if notes is not None:
+            raise NotImplementedError(
+                """
+                Notes are not supported by the odswriter engine,
+                see https://github.com/mmulqueen/odswriter
```
Comment (asishm) [inline pandas/io/excel/_odswriter.py:120]: I don't understand the reference to the linked repo. pandas uses https://github.com/eea/odfpy for writing ODS files.

Comment (diogomsmiranda) [reply pandas/io/excel/_odswriter.py:120]: Hey, I might be wrong but I believe we went to pandas documentation when searching for all the engines pandas currently uses and found this for ExcelWriters, in here explicitly links https://github.com/mmulqueen/odswriter as the engine to write ods files. 

https://pandas.pydata.org/docs/reference/api/pandas.ExcelWriter.html,

Comment (Dacops) [reply pandas/io/excel/_odswriter.py:120]: yes, `odfpy` is indeed used but for reading Open-Document type files (https://pandas.pydata.org/docs/reference/api/pandas.read_excel.html#pandas-read-excel). The writing operation is taken care by that `odswriter` as my colleague referred.

Comment (asishm) [reply pandas/io/excel/_odswriter.py:120]: Thanks @diogomsmiranda @Dacops. I can see why now. The docstring in `pd.ExcelWriter` is incorrect (looks like was added in #47568).

1. If you check the `engine_kwargs` section, it mentions `odf.opendocument.OpenDocumentSpreadsheet(**engine_kwargs)`. None of this construct is available in the linked `odswriter` library, but is available in the `odfpy` library instead https://github.com/eea/odfpy/blob/574f0fafad73a15a5b11b115d94821623274b4b0/odf/opendocument.py#L842.

2. If you check what extra libraries pandas installs when you run `pip install pandas[excel]`, you don't see odswriter listed there. https://github.com/pandas-dev/pandas/blob/dd87dd3ef661ac06e30adc55d25a1f03deda3abf/pyproject.toml#L69

cc @rhshadrach

Comment (diogomsmiranda) [reply pandas/io/excel/_odswriter.py:120]: Oh I see, I'm glad it could be useful to find the mistake. We'll change the link to the correct one, or even better check if `odfpy` has compatibility with notes, and adjust accordingly.

Comment (diogomsmiranda) [reply pandas/io/excel/_odswriter.py]: After looking through the codebase for mentions of comments or notes in cells I didn't find anything, although I did find a couple discussion about the current state of the project, mentioned as "dead".

In the mean time I've fixed the exception to link the right library.

Comment (asishm) [reply pandas/io/excel/_odswriter.py]: > current state of the project, mentioned as "dead".

I think that's fair. The last commit on that repo seems to be 3 years ago.

File: pandas/io/excel/_openpyxl.py (hunk 3 above)
Comment (rhshadrach) [reply pandas/io/excel/_openpyxl.py:496]: My understanding is that `notes_col is None` is only true on the first pass of this for loop, and so this is just a way of getting the first cell in `cells`. The code `next(iter(cells))` also gives you the first cell. The full code would be:

    notes_col = start_col + next(iter(cells)).col + 1

I think this can go below the `if notes is None or notes_col is None:` as we don't need it unless we get passed that.

Comment (Dacops) [reply pandas/io/excel/_openpyxl.py]: got it, my problem was that cells is already an iterator, `next(cells)` is enough. This is fixed now.

Comment (diogomsmiranda) [reply pandas/io/excel/_openpyxl.py]: We opted with doing

```
if notes and cells:
         notes_col = startcol + next(cells).col + 1
```

This works because cells is already an Iterable[ExcelCell] so there's no need for iter(cells).
Plus there's also no need to check if notes_col is None since the checks comes right after the initialisation of notes_col.

**Comment by attack68** [comment]
This looks a lot cleaner now, more targeted and the tests are written to detect the notes. Is there much outstanding other than fixing any failing tests?

**Comment by Dacops** [comment]
> This looks a lot cleaner now, more targeted and the tests are written to detect the notes. Is there much outstanding other than fixing any failing tests?

Nope, I believe this is done, the failing tests were something that was failing globally a while back (32bit test suites were suffering from overflows) but I believe this is fixed now. I'll rebase the PR to address the conflict at the `whatsnew` file and it should pass everything.

**Comment by diogomsmiranda** [comment]
Hey, @rhshadrach for some reason Github is still showing that you requested changes, but I believe we already addressed them. Can you confirm everything is ok?

**Comment by rhshadrach** [comment]
> Can you confirm everything is ok?

I should have some time  in the next few days. As a side note, this would be easier on my end if there wasn't force-pushes involved, as then I could just view the diff from the last time I reviewed this PR. However, because there are force-pushes, I need to review the entire PR again.

**Comment by diogomsmiranda** [comment]
> As a side note, this would be easier on my end if there wasn't force-pushes involved, as then I could just view the diff from the last time I reviewed this PR. However, because there are force-pushes, I need to review the entire PR again.

Thanks for the feedback and sorry for the inconvenience, it's noted for any future contributions.

**Review by attack68 — APPROVED** [review]
Looks well structured and good to me.

**Comment by attack68** [comment]
@rhshadrach can you take a second look when possible to avoid this stalling whilst its green?

File: pandas/io/excel/_odswriter.py [hunk 7]
```diff
@@ -110,6 +113,14 @@ def _write_cells(
         )
         from odf.text import P
 
+        if notes is not None:
+            raise NotImplementedError(
+                """
+                Notes are not supported by the odswriter engine,
+                see https://github.com/eea/odfpy
+                """
```
Comment (rhshadrach) [inline pandas/io/excel/_odswriter.py:121]: This should be a regular string, `"Notes are not..."`, not a triple quoted string.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
When I run

```python
notes = DataFrame({"a": 3 * ["note 1"], "b": 3 * ["note 2"]})
df = DataFrame({"a": [1, 1, 2], "b": [3, 4, 5]})
df.style.set_tooltips(notes).to_excel("test.xlsx", sheet_name="A")
```

The index label for column `a` is missing in the Excel sheet (column `b` is there).

In addition, the following modifications are also broken.

- When `df` has MultiIndex (hierarchical) columns
- `startrow` is nonzero
- Specifying columns as a subset of all columns (e.g. `columns=["b"]`)

**Comment by rhshadrach** [comment]
Thanks for the PR, but it appears to have gone stale. If you could address the comments/requests above and we'd be more than happy to reopen!
